File: easy/two_sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        loop through array and have a nested loop goes though all the remaining digits
        
        """
        # O(n**2) Time
        # O(n)
        # for i in range(len(nums)):
        #     for j in range(i + 1, len(nums)):
        #         if nums[i] + nums[j] == target:
        #             return [i,j]
        # return None

        # faster solution O(n), O(2n) space
        # initialize a hash table, loop through array once and for each element, check whether complement is in hash table, if so, return, if not, add it to the hash table with it's index as the value
        myDict = {}
        for index, i in enumerate(nums):
            complement = target - i
            print(i, complement, target)
            if complement in myDict:
                return [myDict[complement], index]
            myDict[i] = index
            

File: easy/two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 * javascript map methods, get, set, has, delete
 * if need access to index, need a for i loop
 */
const twoSum = function(nums, target) {
  const numsMap = new Map()
  for(let i = 0; i < nums.length; i++) {
      let complement = target - nums[i]
      if (numsMap.has(complement)) {
          return [numsMap.get(complement), i]
      }
      numsMap.set(nums[i], i)
  }
  throw new Error('no Two Sum solution')
};
// O(n) time
// O(n) space
